### ISSUE TITLE
Metrics are not flushed when lambda handler throws an exception

### ISSUE THREAD

**Issue by qtdzz** [awslabs/aws-lambda-powertools-java#285]
**What were you trying to accomplish?**
I am trying to publish metrics even when the lambda handler fails. Apparently, metrics won't be flushed if the handler throws exception: https://github.com/awslabs/aws-lambda-powertools-java/blob/master/powertools-metrics/src/main/java/software/amazon/lambda/powertools/metrics/internal/LambdaMetricsAspect.java#L62

## Expected Behavior

Metrics are flushed regardless of the handler execution status

## Current Behavior

Metrics are flushed only when the handler finishes successfully

## Possible Solution

Some thing likes the following snippet in [the around method](https://github.com/awslabs/aws-lambda-powertools-java/blob/master/powertools-metrics/src/main/java/software/amazon/lambda/powertools/metrics/internal/LambdaMetricsAspect.java#L40)
```
public Object around(ProceedingJoinPoint pjp,
                         Metrics metrics) throws Throwable {

...[omitted code]...

            try {
                Object proceed;
                try {
                     proceed = pjp.proceed(proceedArgs);
                } finally {
                    coldStartDone();
                    validateBeforeFlushingMetrics(metrics);
                    logger.flush();
                }
                return proceed;

            } finally {
                refreshMetricsContext();
            }
}
```

## Steps to Reproduce (for bugs)

1. Add the following class  to `src/test/java/software/amazon/lambda/powertools/metrics/handlers`
```
public class PowertoolsMetricsWithExceptionInHandler implements RequestHandler<Object, Object> {

    @Override
    @Metrics(namespace = "ExampleApplication", service = "booking")
    public Object handleRequest(Object input, Context context) {
        MetricsLogger metricsLogger = metricsLogger();
        metricsLogger.putMetric("CoolMetric", 1);
        throw new IllegalStateException("Whoops, unexpected exception");
    }
}
```
2. Add the following test to `software.amazon.lambda.powertools.metrics.internal.LambdaMetricsAspectTest`
```
    @Test
    public void metricsPublishedEvenHandlerThrowsException() {
        requestHandler = new PowertoolsMetricsWithExceptionInHandler();
        try (MockedStatic<SystemWrapper> mocked = mockStatic(SystemWrapper.class)) {
            mocked.when(() -> SystemWrapper.getenv("AWS_EMF_ENVIRONMENT")).thenReturn("Lambda");
            assertThatExceptionOfType(IllegalStateException.class)
                .isThrownBy(() -> requestHandler.handleRequest("input", context))
                .withMessage("Whoops, unexpected exception");

            assertThat(out.toString())
                .satisfies(s -> {
                    Map<String, Object> logAsJson = readAsJson(s);
                    assertThat(logAsJson)
                        .containsEntry("CoolMetric", 1.0)
                        .containsEntry("Service", "booking")
                        .containsKey("_aws");
                });
        }
    }
```
3. Run unit tests
4. The unit test would fail

## Environment

* **Powertools version used**: master branch
* **Packaging format (Layers, Maven/Gradle)**: Maven
* **AWS Lambda function runtime:** Java 8
* **Debugging logs**

> [How to enable debug mode](https://awslabs.github.io/aws-lambda-powertools-java/#debug-mode)**

```text
# paste logs here
```

**Comment by pankajagrawal16** [comment]
Hey @qtdzz , Thanks for opening an issue with detailed info.

As far as I can remember, this was done on purpose to not capture metrics during exception. Can't exactly remember the reasoning behind that. That being said, I quickly checked the behavior for [aws-lambda-powertools-python](https://github.com/awslabs/aws-lambda-powertools-python) as well and it turns out, it captures metrics regardless of exceptions. Which means, it only makes sense to keep the behavior consistent here as well and hence fix it.

Since you have done most of the researching already, I will be more than happy accept a PR for this, If you wish to :)

If not, I will to raise a fix later this week or next. Thanks again.

**Comment by pankajagrawal16** [comment]
Available in [1.3.0](https://github.com/awslabs/aws-lambda-powertools-java/releases/tag/v1.3.0) now
